### DISCUSSION TITLE
How to disable logging

### DISCUSSION THREAD

**Discussion by 93578237** [abhinavsingh/proxy.py#1195] [Q&A] [answered]
The proposed solution of https://github.com/abhinavsingh/proxy.py/issues/1089 is to add the custom plugin, which prints access logs. In https://github.com/pypa/pip/pull/11091  windows CI has been worked 6 hours and then failed. I think, I need to disable logging at all, but there is no way to do it.

**Comment by abhinavsingh** [comment, marked as answer]
@q0w You can remove `super()` call and it will automatically disable logging from the core, leaving only the `print` statements as the only logging statement for the instance.  https://github.com/pypa/pip/pull/11091/files#diff-35c83feabce838957ec0526f0280604f479df39c03e56ca51b65ed34de098c81R22

However, it's unclear to me why a call to `super()` results in test failures, specially for `windows/1` only runs.  Does your test processes stdout/stderr for assertions?  I looked at the runs but didn't see any specific error logs.  I also see `windows/2` passes.  Wondering if `windows/2` runs is also using proxy.py?

Please let me know if removing `super()` helps.

**Reply by 93578237** [reply]
With removed super() it fails on Windows 3.7 https://github.com/pypa/pip/runs/6904891927?check_suite_focus=true

**Reply by 93578237** [reply]
Yes, I need to check proxy stdout to determine to what proxy pip is connected.

**Reply by abhinavsingh** [reply]
I don't think logging has anything to do with these failures on `Windows`.   You can cancel the workflow instead of letting to run to auto-cancelled.  You might be able to see some stack traces.

Core issue is related to joining the acceptors during test teardown.  I have added a timeout of 1 second, instead of waiting indefinitely.  Looks like joining multi-processing acceptors has an issue on Windows, or at-least is flaky.  In long term, I'll think about a better fix.

Meanwhile, https://test.pypi.org/project/proxy.py/2.4.4.dev3 is out.  I have started a release workflow for `v2.4.4` which will contain this fix.  I hope this will help you get past these Windows hung up test case scenarios. Best

**Reply by abhinavsingh** [reply]
> Yes, I need to check proxy stdout to determine to what proxy pip is connected.

Thank you for this context.  In that scenario, you can continue using your custom plugin to disable `logging` statements.  If you wish, you can also remove `print` statements and simply let `proxy.py` logs vanish in thin air.

But overall, you must also update to using `v2.4.4.dev3` (from test.pypi.org) or `v2.4.4` (from pypi.org).  Release workflow is currently in action, ref https://github.com/abhinavsingh/proxy.py/actions/runs/2504346031

**Reply by abhinavsingh** [reply]
My other advise would be to only start a single acceptor to avoid Windows pitfalls e.g. `with proxy.Proxy(port=8888, num_acceptors=1)`.  Let me know if this alone helps without upgrading to 2.4.4*

**Reply by 93578237** [reply]
>num_acceptors=1

Unfortunately no https://github.com/pypa/pip/runs/6906570097?check_suite_focus=true

**Reply by abhinavsingh** [reply]
> TL;DR -- Please use `subprocess` based strategy which will remove the flakiness/failures happening in Windows environment.

## Alternate Way using Subprocess

In the meantime, I'll recommend you to try [this approach](https://github.com/abhinavsingh/proxy.py/blob/develop/tests/integration/test_integration.py#L165) which cannot be impacted by Windows related join issues and your PR workflows should work fine.  This approach uses subprocess to start proxy, instead of embedded mode.  More importantly, with this approach, during test teardown, your tests need not wait on acceptor joins.  Cleanup of orphaned processes is delegated to GHA automatically.

In-fact, `proxy_py_subprocess` is a `pytest` fixture, if you can, you may directly use this fixture in your PR, or extend on top of it.

Let me know if this helps.

## About bug in Embedded mode

At this point, I can certainly say, this is flaky on Windows _(for `proxy.py` tests itself)_ and looks like it never runs for your PR.  I say so, because after [I merged `timeout=1` PR](https://github.com/abhinavsingh/proxy.py/pull/1196) and ran the release job, workflow got stuck for `Windows + Python 3.10` :(.   Which means `timeout=1` is likely not the right fix here, or at-least not the only fix.

To debug this further, I sent out this PR https://github.com/abhinavsingh/proxy.py/pull/1199 where I am using `acceptor=1`.  I also reverted the `timeout=1` to check if that even was necessary.   But looks like `num_acceptor=1` ain't working for you either.

Because the workflow gets stuck, I believe issue is related to `join`.  There are more join in the code, I only resolved the acceptor ones.  Unsure if one of those are an issue here.  It would have been really helpful, if you could, possibly, run the workflow with debug level logging.  That will help me get more context.

Further on this, following needs to be tried for deeper inspection:
1) Get a Windows 10 VirtualBox setup locally
2) Try to reproduce by running similar usage pattern multiple times
3) If unable to reliably reproduce still, try to clone ur PR and run that specific test locally on Windows.

**Reply by 93578237** [reply]
Thanks for all your efforts, it was really helpful.
> run the workflow with debug level logging.

Unfortunately I have not necessary permissions to do this.

I will try `proxy_py_subprocess` fixture as you said. Thanks a lot.

**Reply by 93578237** [reply]
~~My bad, `num_acceptors = 1` is fine.~~ I think, it was failed because of `127:` as you mentioned.

**Reply by abhinavsingh** [reply]
Yayyyyyyy!!!!!   Thanks for verifying.  Happy this has worked out.  Just to summarize, it's important to use `num_acceptors=1` on windows to avoid join pitfalls.

I will add this in README so that it remains documented under FAQs.  Best

**Comment by abhinavsingh** [comment]
@q0w So to reproduce I sent out [this PR](https://github.com/abhinavsingh/proxy.py/pull/1196) which tries to have a similar workflow as yours i.e. using `with proxy.Proxy():` context manager.  And looks like there is likely an issue on `Windows + Python 3.9`, because [this check](https://github.com/abhinavsingh/proxy.py/runs/6901515705?check_suite_focus=true) is currently in hung state for almost 2 hours.

It worked for `Windows + Python/3.7,3.8,3.10` but went into limbo for `Python/3.9`.   I think it works fine for `Python 3.6` too, the workflow failed due to a flaky test (which uses openssl underneath).  Re-running for `Python 3.6` will 100% succeed it. Coming back to `Python 3.9` scenario, I am trying to put up a Windows VirtualBox locally so that I can at-least reproduce it locally.  I think running it locally on an actual Windows system will immediately reveal the problems, which somehow are hidden on GHA, because job is terminated/canceled by GH due to timeout, likely resulting in no logs (if any).

I'll convert this into an issue, so that we can track this accordingly.

Thanks for bringing this to my notice.

**Comment by abhinavsingh** [comment]
Meanwhile, wanted to point out, here is [an example of how](https://github.com/abhinavsingh/proxy.py/blob/develop/tests/integration/test_integration.py#L165) we use `proxy.py` ourselves during integration testing.  Here we are using a `subprocess` rather than embedded version.  In case, you don't want to wait upon Python 3.9 fix, you could give this a try.  Best

**Comment by abhinavsingh** [comment]
Tried running on `Windows 11 + Python 3.9`, works, lol.   Is it related to `Windows 10 + Python 3.9` only, I am downloading an ISO as I write this to try.   Or is it just related to GHA somehow.  Will update with my findings.

In this screenshot, full screen terminal is simply running `proxy.py`.  I setup entire Windows traffic to go through local proxy.   Then I also wrote a simple script, equivalent to my PR and also how you are using it in your PR.   That seems to work out fine when run on Windows 11.

<img width="1419" alt="Screen Shot 2022-06-15 at 10 39 40 PM" src="https://user-images.githubusercontent.com/126065/173885781-4860605c-4ac2-44ed-9a89-599ea5a2be46.png">

**Comment by abhinavsingh** [comment]
So, finally, I could reproduce and fix.  Turned out, somehow `acceptor.join()` will hang endlessly on `Windows 10 + Python 3.9` combination.  Join happens during test teardown.  Resulting in an experience which made us wonder, test have hanged.  While in reality, tests likely worked, but join never succeeded.

I added some context in the related PR here https://github.com/abhinavsingh/proxy.py/issues/1197

This will require a release to `pypi`.  In the meantime, you can either try directly from git branch of linked PR.  Workflows seems to be working fine now :). I will merge and cut a new release shortly and update thread once after a new release has been made.

Thanks again for bringing this to my notice.  Best.
